### DISCUSSION TITLE
How can I refresh the SQLAlchemy class's relationship?

### DISCUSSION THREAD

**Discussion by nbansal22** [sqlalchemy/sqlalchemy#11641] [Usage Questions]
I have two classes `Parent` and `Child`. I believe SQLAlchemy is doing some internal memoization of the relationship `children` when it is accessed the first time through `delete_children` method and storing the table name information of the `Child` class.

I have a use case where I am changing the  `__table__.name` of `Child` class dynamically to `new_child_table`. After this dynamic update, the `children` relationship points to the old table name and hasn't been updated. Thus, there is a mismatch between the table name  and the column name `parent_id` of  the `Child` class.

The emitted SQL query looks like
```sql
DELETE FROM child_table where parent_id=new_child_table.parent_id;
```

I want to prevent SQLAlchemy from using this internal memoization and check the table mapping again whenever the `children` property is accessed via `delete_children`.

```python
class Parent(Base):
    __tablename__ = "parent_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    children: Mapped[List["Child"]] = relationship(back_populates="parent")
    def delete_children(self):
       self.children.delete()


class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column(ForeignKey("parent_table.id"))
    parent: Mapped["Parent"] = relationship(back_populates="children")


```

**Comment by CaselIT** [comment]
Hi,

> I have a use case where I am changing the __table__.name of Child class dynamically to new_child_table. After this dynamic update, the children relationship points to the old table name and hasn't been updated. Thus, there is a mismatch between the table name and the column name parent_id of the Child class.

This use case is not supported, so I don't think there is a way of doing that

**Comment by nbansal22** [comment]
Is there a way I can clear SQLAlchemy's internal memoization (i.e. by resetting some attribute) which forces it to recheck the relationship? or delete the Parent and Child registration from the memory and reload these classes?

Is there any mechanism at all that will recreate the relationships with correct table name?

**Reply by CaselIT** [reply]
You can use `session.expire` but you may still end up with wrong data since changing the name of a table is not supported.

> Is there any mechanism at all that will recreate the relationships with correct table name?

I don't think there is. You may need to re-create both parent and child classes.

What is the use case that requires you to change the table name?

**Comment by nbansal22** [comment]
We want to change the tables that we want to write or read to without having to write duplicate model classes. I understand that it a niche use case.

`session.expire()` doesn't work. I've tried it. All it does is expire and refresh attributes on the given instance and doesn't recheck the relationship class's table name. 

Where does SQLAlchemy store this relationship information in the memory? How does it make sure that this process gets triggered only once?

**Reply by CaselIT** [reply]
That information is stored multiple places, most notably into a mapper instance that is generated for each generated mapped class. There is no guarantee that where it's stored will be stable between releases, since this is not a supported use case.

**Comment by nbansal22** [comment]
I need the solution for only test cases and not for production code. I am okay with monkey patching or resetting internal variables and the solution being not stable between releases.

I've tried using `clear_mappers()` (https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.clear_mappers) but then I need to setup the mapper back, otherwise I get an error 
```
sqlalchemy.exc.ArgumentError: Column expression, FROM clause, or other columns clause element expected, got <class 'Parent'>
```
Is there a way I can setup the mapping back to fix this error?

**Reply by zzzeek** [reply]
there is no way to surgically modify mappers unless you are motivated to dive into the source and figure out all the places that need to be updated (we dont have resources here to do all that work for you).  you have to clear all the mappers inside a particular [registry](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.registry), then re-map those classes.     that's how SQLAlchemy's suite works.   if you are interested in testing different kinds of mappings that change, then you have to have a test suite that builds up and tears down mappers.
